swap/src/protocol/alice/execution_setup.rs

```diff
@@ -0,0 +1,91 @@
+use crate::{
+    network::request_response::BUF_SIZE,
+    protocol::{
+        alice::{State0, State3},
+        bob,
+    },
+};
+use anyhow::{Context, Error, Result};
+use libp2p::PeerId;
+use libp2p_async_await::BehaviourOutEvent;
+
+#[derive(Debug)]
+pub enum OutEvent {
+    Done(Result<State3>),
+}
+
+impl From<BehaviourOutEvent<State3, (), anyhow::Error>> for OutEvent {
+    fn from(event: BehaviourOutEvent<State3, (), Error>) -> Self {
+        match event {
+            BehaviourOutEvent::Inbound(_, Ok(State3)) => OutEvent::Done(Ok(State3)),
+            BehaviourOutEvent::Inbound(_, Err(e)) => OutEvent::Done(Err(e)),
+            BehaviourOutEvent::Outbound(..) => unreachable!("Alice only supports inbound"),
+        }
+    }
+}
+
+#[derive(libp2p::NetworkBehaviour)]
+#[behaviour(out_event = "OutEvent", event_process = false)]
+pub struct Behaviour {
+    inner: libp2p_async_await::Behaviour<State3, (), anyhow::Error>,
+}
+
+impl Default for Behaviour {
+    fn default() -> Self {
+        Self {
+            inner: libp2p_async_await::Behaviour::new(b"/execution_setup/1.0.0"),
+        }
+    }
+}
+
+impl Behaviour {
+    fn run(&mut self, bob: PeerId, state0: State0) {
+        self.inner
+            .do_protocol_listener(bob, move |mut substream| async move {
+                let alice_message0 = state0.next_message();
```

Comment:
If this message is not used until you transition to `State1`, why is it produced by `State0` and not `State1`?

Reply:
Because it can be produced by `state0`.

Reply:
Well, just because it can doesn't mean it should.
Like we see in this code, it requires us to produce the message unnecessarily early.

Reply:
I agree with you.